### PR TITLE
Proper hint propagation and group formation, driven by flash example

### PR DESCRIPTION
**Pull request by marnold-ibm**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).
aThanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [x] cleanup

#### What this PR does:
Fixes gaps in hint propagation and group formation for coarse tiling.

Specifically
- Moves hint `propagate_named_dims` and `assign_dim_hints` later, just before coarse_tiling, so there are fewer passes inserting nodes that need to be cleaned up
- Add some helpers for copying metadata and call when needed
- Use hint_id rather than variable name for identifying hints
- Create DimHints even for variables that are not used in the op (broadcast) so that groups are formed properly

Code is now producing a single group for our flash example, and produces this group report:

```
[INFO] [temp_passes] coarse_tile_groups: 1 group(s) formed
  group 0 scopes=[hint_4={'slices': {'B': 1}}, hint_5={'slices': {'H': 1}}, hint_6={'slices': {'Lk': 2}}]:
      buf1  aten=['aten.full.default']  (no tiled dims)
      buf3  aten=['aten.full.default']  (no tiled dims)
      buf5  aten=['aten.mul.Tensor']  tiles=["['H']x1"]
      buf6  aten=['aten.clone.default', 'aten.permute.default']  tiles=["['H']x1", "['Lk']x2"]
      buf8  aten=['aten.mul.Tensor']  tiles=["['H']x1", "['Lk']x2"]
      buf9  aten=['spyre.batched_matmul.default']  tiles=["['H']x1", "['Lk']x2"]
      buf10  aten=['aten.clone.default', 'aten.permute.default']  tiles=["['H']x1", "['Lk']x2"]
      buf11  aten=['aten.amax.default']  tiles=["['H']x1"]
      buf12  aten=['aten.maximum.default']  tiles=["['H']x1"]
      buf13  aten=['aten.sub.Tensor']  tiles=["['H']x1"]
      buf14  aten=['aten.exp.default']  tiles=["['H']x1"]
      buf26  aten=['spyre.restickify.default']  (no tiled dims)
      buf15  aten=['aten.mul.Tensor', 'aten.unsqueeze.default']  tiles=["['H']x1"]
      buf16  aten=['aten.sub.Tensor', 'aten.unsqueeze.default']  tiles=["['H']x1", "['Lk']x2"]
      buf17  aten=['aten.exp.default']  tiles=["['H']x1", "['Lk']x2"]
      buf27  aten=['spyre.restickify.default']  tiles=["['H']x1", "['Lk']x2"]
      buf18  aten=['spyre.batched_matmul.default', 'aten.permute.default']  tiles=["['H']x1"]
      buf28  aten=['spyre.restickify.default']  tiles=["['H']x1"]
      buf19  aten=['aten.add.Tensor']  tiles=["['H']x1"]
      buf21  aten=['aten.full.default']  (no tiled dims)
      buf22  aten=['aten.mul.Tensor']  tiles=["['H']x1"]
      buf23  aten=['aten.sum.dim_IntList']  tiles=["['H']x1"]
      buf24  aten=['aten.add.Tensor']  tiles=["['H']x1"]
  ungrouped: [buf25(no hints)]
```

For this program
```
B, H, Lq, Lk, D = 1, 8, 256, 256, 64
block_size = 128

queries_t = torch.randn(B, H, Lq, D, dtype=torch.float16)
keys_t = torch.randn(B, H, Lk, D, dtype=torch.float16)
values_t = torch.randn(B, H, Lk, D, dtype=torch.float16)


def flash(queries, keys, values):
    scale = 1.0 / math.sqrt(math.sqrt(D))

    with spyre_hint(slices={"B": 1}):
        with spyre_hint(slices={"H": 1}):
            with spyre_hint(slices={"Lk": Lk // block_size}):
                output = torch.zeros_like(queries)
                M = torch.full(
                    (B, H, Lq), float("-inf"), device=queries.device, dtype=torch.float16
                )
                denominator = torch.zeros((B, H, Lq), device=queries.device, dtype=torch.float16)
                keys_T = keys.transpose(-1, -2).contiguous()

                scores = torch.matmul(queries * scale, keys_T * scale)  # B, H, Lq, Lk
                scores = scores.transpose(-1, -2).contiguous()  # avoid stick reduction: B, H, Lk, Lq
                block_max = torch.amax(scores, dim=-2)  # B, H, Lq
                max_running = torch.maximum(M, block_max)  # B, H, Lq

                exp_scores = torch.exp(
                    scores - max_running.unsqueeze(-2)
                )  # B, H, Lk, Lq
                correction = torch.exp(M - max_running)  # B, H, Lq

                denominator = denominator * correction + exp_scores.sum(dim=-2)  # B, H, Lq
                output = (
                    output * correction.unsqueeze(-1)
                    + torch.matmul(exp_scores.transpose(-1, -2), values)  # B, H, Lq, D
                )
                M = max_running

    return output / denominator.unsqueeze(-1)
````

#### Which issue(s) this PR is related to:
https://github.com/torch-spyre/torch-spyre/issues/2150
#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [x] cleanup

#### Which issue(s) this PR is related to:
https://github.com/torch-spyre/torch-spyre/issues/2150

### CODE REVIEW

**Review by dgrove-oss — APPROVED** [review]
Some minor items noted for cleanup.  They can either be addressed in this PR or left for a follow up.


-------

  Summary

  Fixes hint propagation gaps in the coarse-tiling pipeline so ops that straddle restickify injection, mm→bmm rewrites, and scalar-fill (broadcast) lowerings all land in the same coarse-tile group. Groups are now keyed by frozenset(hint_id) rather
  than (split_count, dim_index), and sentinel DimHint entries are created for ops that have no matching loop dimension but are still inside a hint scope.

  Review

  BLOCKERs

  None.

  ---
  SUGGESTIONs
  
  - range_size type annotation is misleading (propagate_hints.py, DimHint).
  The sentinel value 0 means "no real dim" but range_size is still typed as plain int (just wrapped in (int  # ...) parentheses for the comment). Consider range_size: int = 0 as the default for sentinel construction, or at minimum a named constant
  _SENTINEL_RANGE = 0 so callers don't scatter magic zeros. The three-line parenthesized form for a simple inline comment is also stylistically unusual.
  - _find_spec_op falls back to ops[0] silently (temp_passes.py).
  If the entire group somehow consists of all-sentinel ops (unlikely but possible if hints propagated only to scalar-fill ops), ops[0] has range_size=0 everywhere and _group_spec returns []. The resulting coarse_tile(group, []) call may silently
  no-op rather than raising a clear error. A assert spec, ... after _group_spec(...) in hints_to_coarse_tile_groups would catch this earlier.
  - Implicit ops[0] for spec and logging in hints_to_coarse_tile_groups vs. _find_spec_op.
  The group header in the logging block calls _find_spec_op(group_ops) to get hint_descs, which is correct. However, if a group has only sentinel-hint ops (all range_size==0) the hint_descs list is empty and the group header prints group N scopes=[].
  This is confusing during debugging. A small note or fallback would help.
  - convert_constant_with_graph_node hint propagation is a behavior change (temp_passes.py:558).
  The new copy_fx_custom_meta(node, const_node) call will copy hint metadata onto newly-created py_const nodes. Since these constants are outside any user hint scope in practice, this is likely harmless, but it's worth a comment noting why constant
  nodes need the hint copied (i.e., so that get_op_hints on downstream ops that fuse with the constant doesn't lose hint info).
  - resolve_hints (old function) is still present in temp_passes.py alongside the new assign_dim_hints.
  Looking at the current file, resolve_hints is renamed to assign_dim_hints in the diff, but double-check there's no dangling re-export or call site referencing resolve_hints anywhere else (e.g., in __init__.py, passes.py, or tests). passes.py
  already imports the new name correctly.
  - Sentinel DimHint picks only the first key from the hint dict (temp_passes.py, assign_dim_hints).
  When creating the sentinel for an unmatched hint scope, the code does name, count = next(iter(dims.items())). Dict iteration order is insertion-order in Python 3.7+, but dims is reconstructed from hint_dict.get(key) which is a user-supplied dict.
  This is fine in practice but a brief comment noting that the choice of name/count here is only for grouping identity (not for actual tiling) would help the next reader.

  ---
  QUESTIONs
  
  - Consecutive-group assumption: The old grouping logic required consecutive ops with the same hint key to form a group. The new frozenset(hint_id) key preserves this: two ops must have identical hint-ID sets AND be consecutive in the operations
  list to land in the same group. Is this still correct for flash attention, where buf25 ends up ungrouped (as shown in the PR description)? The group report in the PR body shows it as ungrouped: [buf25(no hints)], which looks right — but confirming
  the intent is fine.
  - Interaction with SPYRE_COARSE_TILE_MAX_OP_INDEX: The earlier commit in this branch added that debug knob but the final commit description says it reverted coarse_tile.py to main. Was SPYRE_COARSE_TILE_MAX_OP_INDEX dropped from the final PR, or is
  it still present? The diff doesn't show it in coarse_tile.py, so it appears to be gone — just confirming this is intentional and not accidentally reverted away.

  ---
  Verdict
  
  APPROVE — all CI checks pass, DCO is present on every commit, license headers are correct, and the logic is sound. The suggestions above are quality improvements but none are blocking.
